zaza/utilities/installers.py

```diff
@@ -0,0 +1,571 @@
+    :type name: str
+    :returns: True if the user exists
+    :rtype: bool
+    """
+    cmd = ["grep", "-c", "^{name}:".format(name=name), "/etc/passwd"]
+    try:
+        ssh_fn(cmd)
+    except Exception:
+        return False
+    return True
+
+
+def user_directory(ssh_fn, name=None):
+    """Get the directory for the user.
+
+    The ssh_fn is a function that takes a command and runs it on the remote
+    system.
+
+    If :param:`name` is None, then the user for the ssh_fn is detected,
+    otherwise the home dir or specified user is detected.
+
+    :param ssh_fn: a ssh_fn that can run commands on the unit.
+    :type ssh_fn: Callable[[str], str]
+    :param name: the name of the user to get the directory for.
+    :type name: Option[str]
+    :returns: True if the user exists
+    :rtype: bool
+    """
+    if name is None:
+        cmd = ["echo", "$HOME"]
+    else:
+        cmd = ["echo", "~{name}".format(name=name)]
+    return ssh_fn(cmd).strip()
+
+
+class SystemdControl:
+    """Provide a mechanism to control an daemon as a systemd process.
+
+    The control file will be stored in /etc/systemd/system
+
+    It is assumed to be a simple daemon that can be stopped and started using
+    typical control signals.
+
+
+    """
+
+    SYSTEMD_FILE = textwrap.dedent("""
+        [Unit]
+        Description=SystemdControl file for {name}
+
+        [Service]
+        ExecStart={exec_start}
+        User={name}
+        Environment=PYTHONUNBUFFERED=1
+        Restart=on-failure
+        RestartSec=5s
+
+        [Install]
+        WantedBy=default.target
+        """)
+
+    def __init__(
+            self, ssh_fn, scp_fn, name, execute):
+        """Initialise a SystemdControl object.
+
+        :param ssh_fn: a function that runs commands on the instance; this
+            needs to have sudo access.
+        :type ssh_fn: Callable[[str], str]
+        :param scp_fn: a function that can copy files to the instance.
+        :type scp_fn: Callable[[str, str, bool], None]
+        :param name: name of the unit (short, no spaces)
+        :type name: str
+        :param execute: the command+options that the systemd unit will execute
+            to run the program.
+        :type execute: str
+        """
+        self.ssh_fn = ssh_fn
+        self.scp_fn = scp_fn
+        self.name = name
+        self.execute = execute
+        self._enabled = False
+        self._installed = False
+        self._home_var = None
+        self._stopped = True
+
+    @property
+    def _systemd_dir(self):
+        return "/etc/systemd/system"
+
+    @property
+    def _systemd_filename(self):
+        return "{}.service".format(self.name)
+
+    @property
+    def _systemd_file(self):
+        return os.path.join(self._systemd_dir, self._systemd_filename)
+
+    @property
+    def _home(self):
+        if self._home_var is not None:
+            return self._home_var
+        self._home_var = self.ssh_fn("echo $HOME").strip()
```

Comment:
Isn't this duplicating code from the user_directory function defined earlier?

Reply:
Good spot!  It's funny when writing this stuff how things like this get away from you.  It just shows the absolutely utility and benefit of reviews.  Thanks.